### DISCUSSION TITLE
Can't set metadata for gfm output with lua filter

### DISCUSSION THREAD

**Discussion by qiushiyan** [quarto-dev/quarto-cli#10668] [Q&A] [answered]
### Description

I am creating a lua filter in Quarto 1.5 to extract all headings from the source document, and want to add a `headings` field to the output gfm document.

Here is my lua filter and qmd 

```lua
-- Function to create a slug from a string
local function slugify(s)
    return s:lower()
        :gsub("[^%w%s-]", "") -- remove non-word chars
        :gsub("%s+", "-")     -- replace spaces with hyphens
        :gsub("^-+", "")      -- trim starting hyphens
        :gsub("-+$", "")      -- trim ending hyphens
end

-- Table to store extracted headings
local headings = {}

-- Function to process headers
function Header(el)
    if el.level == 2 or el.level == 3 then
        local title = pandoc.utils.stringify(el.content)
        table.insert(headings, {
            depth = el.level,
            title = title,
            slug = slugify(title)
        })
    end
    return el
end

function Meta(meta)
    quarto.log.output("headings: " .. pandoc.utils.stringify(headings))
    meta["headings"] = headings
    return meta
end
```

```qmd
---
format:
  gfm:
    variant: +yaml_metadata_block
    filters:
      - filters/headings.lua
---

# Main Title

## First Section

Some content here.

### Subsection A

More content.

## Second Section

Even more content.

### Subsection B

Final content.
```

During render, the filter is getting headings correctly and prints

```
(headings: 2first-sectionFirst Section3subsection-aSubsection A2second-sectionSecond Section3subsection-bSubsection B)
```

But no `headings` field is added to the output.

**Comment by mcanouil** [comment]
Why would you think there will be something? You are adding the headings as metadata not as output.

**Reply by qiushiyan** [reply]
How could I output the updated metadata? I assumed the output will contain the updated meta object. I found this example on the pandoc lua documentation to update the date field and I thought this should work in a similar way

```lua
function Meta(m)
  if m.date == nil then
    m.date = os.date("%B %e, %Y")
    return m
  end
end
```

**Reply by mcanouil** [reply]
You need to make the metadata part of the main AST content.

```lua
function Pandoc(doc)
  ...
  return doc
end
```

I think there is a format flag from Pandoc to keep the metadatablock but can't reminder it.

```yaml
format: gfm+yaml_metadata_block
```

**Reply by qiushiyan** [reply]
Thanks for the tip, and I am using `gfm+yaml_metadata_block` but I still could not get it working after adding this

```lua
function Pandoc(doc)
    local meta = doc.meta

    meta.headings = headings
    for _, h in ipairs(meta.headings) do
        print(h.title)
    end
    return pandoc.Pandoc(doc.blocks, meta)
end
```

From the print results I can tell that the headings are added correctly to the meta object, it is just that they are not pushed to the content. I also tried loop through the `doc.meta` object and it's empty, there is no field such as `title` or `author`. Perhaps this is a problem of how the gfm format deals with the meta object?

**Comment by cderv** [comment, marked as answer]
Your filter is perfectly fine, as it correspond to the orders filter are run (first Header will be processed, before any Meta)

If you try with pandoc it will work (removing `quarto.log.output` specific API)
```
quarto pandoc index.qmd -t gfm+yaml_metadata_block -L headings.lua -s
```

However, with Quarto, we don't support this use case for now. For some reason, we are taking over `+yaml_metadata_block` behavior to recreate the feature, but without taking into account any modification by user in Lua filter. 

To be clear, very simply, we do get the YAML in the input, and put it back on the output when asked. So currently doing what you want is not possible. We would need to reimplement the logic differently. 

That is a current limitation of Quarto

**Reply by cderv** [reply]
FWIW, I opened an issue to track the idea for our future development
- https://github.com/quarto-dev/quarto-cli/issues/10686

**Reply by qiushiyan** [reply]
Thanks for the explanation.
